org.eclipse.lsp4e/src/org/eclipse/lsp4e/internal/CharsInputStream.java

```diff
@@ -0,0 +1,276 @@
+	 * @return {@code true} if there are still bytes left in the {@link #byteBuffer}
+	 *         after flushing, or if the encoder still has more bytes to flush;
+	 *         {@code false} if the flush is complete and no bytes remain.
+	 */
+	private boolean flushEncoder() throws IOException {
+		if (encoderState == EncoderState.DONE)
+			return false;
+
+		if (encoderState == EncoderState.ENCODING) {
+			encoderState = EncoderState.FLUSHING;
+		}
+
+		// flush
+		byteBuffer.clear();
+		final CoderResult result = encoder.flush(byteBuffer);
+		byteBuffer.flip();
+
+		if (result.isOverflow()) {
+			// the byteBuffer has been filled, but there are more bytes to be flushed.
+			// after reading all available bytes from byteBuffer, flushEncoder() needs to
+			// be called again to process the remaining data.
+			return true;
+		}
+
+		if (result.isError()) {
+			result.throwException();
+		}
+
+		encoderState = EncoderState.DONE;
+		return byteBuffer.hasRemaining();
+	}
+
+	@Override
+	public int read() throws IOException {
+		if (!byteBuffer.hasRemaining() && !refillByteBuffer())
+			return EOF;
+		return byteBuffer.get() & 0xFF; // next byte as an unsigned integer (0 to 255)
+	}
+
+	@Override
+	public int read(final byte[] buf, final int off, final int bytesToRead) throws IOException {
+		Objects.checkFromIndexSize(off, bytesToRead, buf.length);
+		if (bytesToRead == 0)
+			return 0;
+
+		int bytesRead = 0;
+		int bytesReadable = byteBuffer.remaining();
+
+		while (bytesRead < bytesToRead) {
+			if (bytesReadable == 0) {
+				if (refillByteBuffer()) {
+					bytesReadable = byteBuffer.remaining();
+				} else
+					return bytesRead == 0 ? EOF : bytesRead;
+			}
+
+			final int bytesToReadNow = Math.min(bytesToRead - bytesRead, bytesReadable);
+			byteBuffer.get(buf, off + bytesRead, bytesToReadNow);
+			bytesRead += bytesToReadNow;
+			bytesReadable -= bytesToReadNow;
+		}
+
+		return bytesRead;
+	}
+
+	/**
+	 * Refills the {@link #byteBuffer} by reading characters from the character
+	 * supplier, encoding them, and storing the resulting bytes into the
+	 * {@link #byteBuffer}.
+	 *
+	 * @return {@code true} if the buffer was successfully refilled and has bytes
+	 *         available for reading, {@code false} if the end of the stream is
+	 *         reached and there are no more bytes to read.
+	 */
+	private boolean refillByteBuffer() throws IOException {
+		if (encoderState == EncoderState.DONE)
+			return false;
+
+		if (encoderState == EncoderState.FLUSHING)
+			return flushEncoder();
+
+		final int charsLen = charsLength.getAsInt();
+
+		// if EOF is reached transition to flushing
+		if (charIndex >= charsLen) {
+			// finalize encoding before switching to flushing
+			byteBuffer.clear();
+			final CoderResult result = encoder.encode(CharBuffer.allocate(0), byteBuffer, true /* signal EOF */);
+			byteBuffer.flip();
+			if (result.isError()) {
+				result.throwException();
+			}
+			return flushEncoder();
+		}
+
+		try {
+			charBuffer.clear();
```

Comment:
should we not do the `charBuffer.clear` rather at the end of the method in a `finally` block so that we do not hold data that will never be used until the next call to refill?

Reply:
the charBuffer contains a fixed size char[] array. I don't see any advantage clearing the buffer at the end. I prefer clearing the buffer just before I starting refilling and using it.